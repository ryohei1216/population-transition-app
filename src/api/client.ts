type EmptyObject = Record<string, never>;
// TODO：エンドポイントを環境変数へ
const PATH_PREFIX = 'https://opendata.resas-portal.go.jp';

export default async function get<
  Query extends { [key: string]: string | number } = EmptyObject,
  Response = EmptyObject
>(path: string, query: Query): Promise<Response> {
  const url = new URL(PATH_PREFIX + path);
  Object.entries({
    ...query,
  })
    .filter(([key, value]) => value !== undefined)
    .forEach(([key, value]) => {
      url.searchParams.append(key, String(value));
    });

  const res = await fetch(url.toString(), {
    method: 'GET',
    headers: {
      // TODO:API-KEYを環境変数へ
      'X-API-KEY': 'CX1V0FJ7UDeE2vnwCGDuFMexbADUAYuGAfIEnTlj',
    },
  });

  // eslint-disable-next-line
  return await res.json();
}
