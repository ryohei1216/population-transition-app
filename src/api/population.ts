import get from './client';

export type PerPopulation = {
  label: string;
  data: {
    year: number;
    value: number;
  }[];
};
export type TotalPopulation = {
  message: null;
  result: {
    boundaryYear: number;
    data: PerPopulation[];
  };
};

export type Query = {
  prefCode: number;
  cityCode: string;
};
export async function getPopulation(query: Query): Promise<TotalPopulation> {
  const res = await get<Query, TotalPopulation>(
    '/api/v1/population/composition/perYear',
    {
      prefCode: query.prefCode,
      cityCode: query.cityCode,
    }
  );
  return res;
}
