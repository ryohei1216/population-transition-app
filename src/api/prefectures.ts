import get from './client';

export type Prefecture = {
  prefCode: number;
  prefName: string;
};

export type Prefectures = {
  message: null;
  result: Prefecture[];
};

export async function getPrefectures() {
  const res = await get<Record<string, never>, Prefectures>(
    '/api/v1/prefectures',
    {}
  );
  return res;
}
