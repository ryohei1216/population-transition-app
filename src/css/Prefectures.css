.prefectures-container {
  display: flex;
  width: 100%;
  max-width: 900px;
  margin: 0 auto;
  padding: 0 40px 40px 40px;
  flex-wrap: wrap;
}
.checkbox-wrapper {
  width: 100px;
  font-size: 16px;
  align-items: center;
  text-align: left;
}
