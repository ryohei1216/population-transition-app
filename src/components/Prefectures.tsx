import React, { FC, useState } from 'react';
import { Prefecture } from '../api/prefectures';
import usePrefectures from '../hooks/usePrefecture';

type Props = {
  setPrefectureList: React.Dispatch<React.SetStateAction<Prefecture[]>>;
};
const Prefectures: FC<Props> = ({ setPrefectureList }) => {
  const { prefectures } = usePrefectures();

  const onClickCheckbox = (prefecture: Prefecture) => {
    setPrefectureList((preList) => {
      if (preList.some((el) => el.prefCode === prefecture.prefCode)) {
        const newArray = preList.filter(
          (el) => el.prefCode !== prefecture.prefCode
        );
        return newArray;
      }

      return [...preList, prefecture];
    });
  };
  return (
    <div>
      <h3>都道府県</h3>
      <div className="prefectures-wrapper">
        <div className="prefectures-group">
          {prefectures?.result &&
            prefectures.result.map((prefecture) => (
              <div className="checkbox-wrapper" key={prefecture.prefCode}>
                <input
                  data-testid={prefecture.prefName}
                  type="checkbox"
                  onClick={() => onClickCheckbox(prefecture)}
                />
                {prefecture.prefName}
              </div>
            ))}
        </div>
      </div>
    </div>
  );
};

export default Prefectures;
