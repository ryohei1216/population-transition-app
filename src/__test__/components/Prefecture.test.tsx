import React, { useState } from 'react';
import {
  render,
  screen,
  cleanup,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { Prefecture } from '../../api/prefectures';
import 'whatwg-fetch';
import Prefectures from '../../components/Prefectures';

/**
 * テスト内容
 * １．スナップショットテスト ２．ユニットテスト
 * APIで都道府県の取得ができてるか
 * チェックボックスを１回目と２回め押すときに表示が変わるか
 * 押したときに関数が呼ばれているか
 */

const server = setupServer(
  rest.get(
    `https://opendata.resas-portal.go.jp/api/v1/prefectures`,
    (req, res, ctx) => {
      return res(
        ctx.status(200),
        ctx.json({
          message: null,
          result: [
            {
              prefCode: 1,
              prefName: '北海道',
            },
            {
              prefCode: 2,
              prefName: '青森県',
            },
          ],
        })
      );
    }
  )
);

beforeAll(() => server.listen());
afterEach(() => {
  server.resetHandlers();
  cleanup();
});
afterAll(() => server.close());

describe('Prefectureコンポーネントのテスト', () => {
  test('APIから都道府県の取得・表示', async () => {
    const setPrefectureList = jest.fn();
    const { asFragment } = render(
      <Prefectures setPrefectureList={setPrefectureList} />
    );
    expect(asFragment()).toMatchSnapshot();

    expect(await screen.findByTestId('1')).toHaveTextContent('北海道');
    expect(await screen.findByTestId('2')).toHaveTextContent('青森県');

    expect(asFragment()).toMatchSnapshot();
  });

  test('checkboxクリック', async () => {
    const setPrefectureList = jest.fn();
    const { asFragment } = render(
      <Prefectures setPrefectureList={setPrefectureList} />
    );
    expect(asFragment()).toMatchSnapshot();

    const checkbox1 = await screen.findByTestId('北海道');
    fireEvent.click(checkbox1);
    expect(setPrefectureList).toHaveBeenCalled();
    expect(checkbox1).toBeChecked();

    fireEvent.click(checkbox1);
    expect(setPrefectureList).toHaveBeenCalled();
    expect(checkbox1).not.toBeChecked();

    expect(asFragment()).toMatchSnapshot();
  });
});
