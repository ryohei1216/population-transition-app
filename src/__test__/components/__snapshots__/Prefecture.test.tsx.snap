// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prefectureコンポーネントのテスト APIから都道府県の取得・表示 1`] = `
<DocumentFragment>
  <div>
    <h3>
      都道府県
    </h3>
    <div
      class="prefectures-wrapper"
    >
      <div
        class="prefectures-group"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Prefectureコンポーネントのテスト APIから都道府県の取得・表示 2`] = `
<DocumentFragment>
  <div>
    <h3>
      都道府県
    </h3>
    <div
      class="prefectures-wrapper"
    >
      <div
        class="prefectures-group"
      >
        <div
          class="checkbox-wrapper"
          data-testid="1"
        >
          <input
            data-testid="北海道"
            type="checkbox"
          />
          北海道
        </div>
        <div
          class="checkbox-wrapper"
          data-testid="2"
        >
          <input
            data-testid="青森県"
            type="checkbox"
          />
          青森県
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Prefectureコンポーネントのテスト checkboxクリック 1`] = `
<DocumentFragment>
  <div>
    <h3>
      都道府県
    </h3>
    <div
      class="prefectures-wrapper"
    >
      <div
        class="prefectures-group"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Prefectureコンポーネントのテスト checkboxクリック 2`] = `
<DocumentFragment>
  <div>
    <h3>
      都道府県
    </h3>
    <div
      class="prefectures-wrapper"
    >
      <div
        class="prefectures-group"
      >
        <div
          class="checkbox-wrapper"
          data-testid="1"
        >
          <input
            data-testid="北海道"
            type="checkbox"
          />
          北海道
        </div>
        <div
          class="checkbox-wrapper"
          data-testid="2"
        >
          <input
            data-testid="青森県"
            type="checkbox"
          />
          青森県
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
